packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,683 @@
+          } else if (i == boxes.length - 1) {
+            rChildRect = RRect.fromRectAndCorners(childRect,
+                topRight: const Radius.circular(3.0),
+                bottomRight: const Radius.circular(3.0));
+          } else {
+            rChildRect = RRect.fromRectAndCorners(childRect);
+          }
+          _rects.add(rChildRect);
+          start += childWidth;
+        }
+        break;
+    }
+
+    size = constraints.constrain(overallSize);
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    fillPaint.clear();
+    for (int i = 0; i < boxes.length; i += 1) {
+      if (textDirection == TextDirection.ltr) {
+        _paintChild(context, offset, boxes.elementAt(i), i, i == 0,
+            i == boxes.length - 1);
+      } else {
+        _paintChild(context, offset, boxes.elementAt(i), i,
+            i == boxes.length - 1, i == 0);
+      }
+    }
+  }
+
+  void _paintChild(PaintingContext context, Offset offset, RenderBox child,
+      int slot, bool isFirst, bool isLast) {
+    if (child == null) {
+      return;
+    }
+
+    final Rect childRect = _boxRect(child).shift(offset);
+    RRect childRRect;
+    if (isFirst) {
+      childRRect = RRect.fromRectAndCorners(childRect,
+          topLeft: const Radius.circular(3.0),
+          bottomLeft: const Radius.circular(3.0));
+    } else if (isLast) {
+      childRRect = RRect.fromRectAndCorners(childRect,
+          topRight: const Radius.circular(3.0),
+          bottomRight: const Radius.circular(3.0));
+    } else {
+      childRRect = RRect.fromRectAndCorners(childRect);
+    }
+
+    Color color = CupertinoColors.white;
+    if (selectedValue != null && selectedValue.value == slot) {
+      color = CupertinoColors.activeBlue;
+    } else if (pressedValue != null && pressedValue.value == slot) {
+      color = _kPressedBackground;
+    }
+
+    if (textDirection == TextDirection.ltr) {
+      fillPaint.add(new Paint()
+        ..color = color
+        ..style = PaintingStyle.fill);
+    } else {
+      fillPaint.insert(0, new Paint()
+        ..color = color
+        ..style = PaintingStyle.fill);
+    }
+
+    context.canvas.drawRRect(
+        childRRect,
+        textDirection == TextDirection.ltr ? fillPaint.last : fillPaint.first);
+    context.canvas.drawRRect(
+        childRRect,
+        new Paint()
+          ..color = CupertinoColors.activeBlue
+          ..strokeWidth = 2.0
+          ..style = PaintingStyle.stroke);
+    context.paintChild(child, _boxParentData(child).offset + offset);
+  }
+
+  @override
+  bool hitTestChildren(HitTestResult result, {@required Offset position}) {
+    assert(position != null);
+    for (int i = 0; i < _children.length; i += 1) {
+      if (_rects.elementAt(i).contains(position)) {
+        _hitTestLog.add(children.elementAt(i).key);
+        return _children.elementAt(i).hitTest(result,
+            position: (Offset.zero & _boxSize(_children.elementAt(i))).center);
```

Comment:
for things like this use the built-in features (e.g. in this case `defaultHitTestChildren`)